sbi/samplers/vi/vi_divergence_optimizers.py

```diff
@@ -0,0 +1,710 @@
+
+    def __init__(
+        self,
+        potential_fn: BasePotential,
+        q: Distribution,
+        n_particles: int = 256,
+        clip_value: float = 5.0,
+        optimizer: "Optimizer" = Adam,
+        scheduler: object = ExponentialLR,
+        eps: float = 1e-5,
+        **kwargs,
+    ):
+        """This is a wrapper around a PyTorch optimizer which is used to minimize some
+         loss for variational inference.
+
+        Args:
+            potential_fn: Potential function of the target i.e. the posterior density up
+                to normalization constant.
+            q: Variational distribution
+            n_particles: Number of samples used to estimate gradients.
+            clip_value: Norm value on which gradients are clipped.
+            optimizer: Base class for an pytorch optimizer.
+            scheduler: Base class for an pytorch scheduler.
+            eps: This value determines the sensitivity of the convergence checks.
+            kwargs: All additional arguments associated with optimizer, scheduler such
+                as learning_rate.
+
+
+        """
+
+        self.potential_fn = potential_fn
+        self.q = q
+        self.prior = potential_fn.prior
+        self.device = potential_fn.device
+        self.to(self.device)
+
+        self.n_particles = n_particles
+        self.clip_value = clip_value
+        self.learning_rate = kwargs.get("lr", 1e-3)
+        self.retain_graph = kwargs.get("retain_graph", False)
+        self._kwargs = kwargs
+
+        # This prevents error that would stop optimization.
+        self.q.set_default_validate_args(False)
+        self.prior.set_default_validate_args(False)
+
+        # Manage modules if present.
+        if hasattr(self.q, "modules"):
+            self.modules = nn.ModuleList(self.q.modules())
+        else:
+            self.modules = nn.ModuleList()
+        self.modules.train()
+
+        # Ensure that distribution has parameters and that these are on the right device
+        if not hasattr(self.q, "parameters"):
+            raise ValueError("Your distribution has not parameters please add them!")
+        self.to(self.device)
+
+        # Keep a state to resolve invalid values
+        self.state_dict = [para.data.clone() for para in self.q.parameters()]
+
+        # Init optimizer and scheduler with correct arguments
+        opt_kwargs = filter_kwrags_for_func(optimizer.__init__, kwargs)
+        scheduler_kwargs = filter_kwrags_for_func(scheduler.__init__, kwargs)
+        scheduler_kwargs["gamma"] = scheduler_kwargs.get("gamma", 1.0)
+        self._optimizer = optimizer(self.q.parameters(), **opt_kwargs)
+        self._scheduler = scheduler(self._optimizer, **scheduler_kwargs)
+        self._scheduler._step_count = 2  # Prevents  warning...
+
+        # Loss and summary
+        self.eps = eps
+        self.num_step = 0
+        self.warm_up_was_done = False
+        self.losses = np.ones(2000)
```

Comment:
Any reason to use numpy here?

Reply:
Mhh, I guess no. I think I used some numpy thinks previously, but I remove it. So I guess we can switch to PyTorch.

Reply:
Yup torch would be great!